### PR TITLE
Firestore C++: make FSTDispatchQueue delegate to C++ implementation

### PR DESCRIPTION
**Pull request by var-const**
FSTDispatchQueue now doesn't contain any logic of its own and instead
just passes through all method calls to AsyncQueue (backed by an
ExecutorLibdispatch). Additionally:
* add a `Clear` method to Executor interface that discards pending
  operations if possible and blocks until the rest of the operations
  complete. This is to ensure that libdispatch doesn't try to access
  AsyncQueue after it's destroyed.
* small fixes in FSTDispatchQueue tests.

### CODE REVIEW

**Comment by var-const** [comment]
I'll hold off merging this for a time.

File: Firestore/Source/Core/FSTFirestoreClient.mm [hunk 1]
```diff
@@ -157,6 +157,11 @@ - (instancetype)initWithDatabaseInfo:(const DatabaseInfo &)databaseInfo
   return self;
 }
 
+- (void)dealloc {
+  // Clear all delayed operations so that GCD doesn't try to run them after the queue is destroyed.
+  [self.workerDispatchQueue clear];
```
Comment (wilhuff) [inline Firestore/Source/Core/FSTFirestoreClient.mm]: How does this interact with tasks scheduled into the future? The scenario I'm thinking of is:
  1. User has active listens
  2. Cancels them all
  3. Deletes the Firestore instance.
The problem is that at (2) we kick off an idle timer on the watch stream. There's a parallel problem with the write stream.

File: Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm [hunk 2]
```diff
@@ -51,15 +42,34 @@ void DispatchAsync(const dispatch_queue_t queue, std::function<void()>&& work) {
   });
 }
 
-void DispatchSync(const dispatch_queue_t queue, Executor::Operation work) {
+void DispatchSync(const dispatch_queue_t queue, std::function<void()> work) {
+  FIREBASE_ASSERT_MESSAGE(
+      StringViewFromDispatchLabel(dispatch_queue_get_label(queue)) !=
+          StringViewFromDispatchLabel(
+              dispatch_queue_get_label(dispatch_get_main_queue())),
+      "Calling dispatch_sync on the main queue will lead to a deadlock.");
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm]: `dispatch_sync(dispatch_get_main_queue(), ^{})` works just fine:

```c++
#import <Foundation/Foundation.h>

#include <cstdlib>
#include <iostream>

int main() {
  dispatch_queue_t queue = dispatch_get_main_queue();
  dispatch_queue_t secondary = dispatch_queue_create("secondary", nullptr);
  dispatch_async(secondary, ^{
    dispatch_sync(queue, ^{
      std::cout << "Hello world!\n";
    });
    exit(0);
  });

  dispatch_main();
}
```
What doesn't work is `dispatch_sync` or `dispatch_async` on the main queue if `dispatch_main` hasn't been called yet. That's called normally as a part of the run loop of iOS/macOS applications. Perhaps the problem is that the GoogleTest-based tests don't do that?

File: Firestore/Source/Core/FSTFirestoreClient.mm (hunk 1 above)
Comment (var-const) [reply Firestore/Source/Core/FSTFirestoreClient.mm]: If you don't mind, I'd like to discuss this out of band, because I think it's the most significant one.

The idea is to cancel any delayed operations that might still be on the queue. This is done specifically for the libdispatch implementation. In C++, the queue owns both the worker thread and the operations, so the destructor will just clear the operations container and join the thread.  With libdispatch, it's impossible to cancel any operation once it's dispatched (or clear the queue, or even suspend it -- apparently libdispatch crashes if a queue is being destroyed while it's in suspended state). IIUC, it can't be guaranteed that the dispatch queue gets released in a predictable manner, so my concern is that libdispatch may get to run a delayed operation after Firestore instance is destroyed. I have this local test (don't know if it's possible to do this in a more reasonable manner, without relying on `sleep`):
```cpp
  bool called = false;
  {
    AsyncQueue other_queue{std::unique_ptr<internal::Executor>(GetParam()())};
    other_queue.Enqueue([&] {
      other_queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(300), kTimerId1,
                                    [&called] { called = true; });
    });
    other_queue.Clear(); // Comment or uncomment
  }
  std::this_thread::sleep_for(std::chrono::seconds(1));
  EXPECT_FALSE(called);
```
With a call to `Clear`, everything works fine. Without a call to `Clear`, there's a data race when libdispatch gets to run the operation after `other_queue` is destroyed. The way I wrote `AsyncQueue` checks is that the lambda passed to `EnqueueAfterDelay` is wrapped into another lambda that contains a reference to the `AsyncQueue` (this is to avoid `Executor` being aware of the `AsyncQueue` logic). Even if that were not the case, the `DelayedOperation` will try to remove itself from the `Executor`, which is also destroyed at this point. Even if that was somehow avoided, I suspect that any objects that the operation references are probably gone by this point.

What are your thoughts on this? Can Firestore instance be destroyed and recreated during the lifetime of the app? If this can only happen on shutdown, perhaps we don't care? (but it seems it may sometimes lead to a crash on access to deallocated memory, so crashing on shutdown still seems like a problem).

For the scenario you describe, `clear` will cancel the idle timer. Sorry, I presumed  streams don't outlive the Firestore instance, sounds like it's not the case? Can you explain in more detail how they interact? Can Firestore instance be recreated?

Comment (var-const) [reply Firestore/Source/Core/FSTFirestoreClient.mm]: To summarize our offline discussion:
* the summary is that `clear` should not be needed;
* classes that schedule delayed operations are already supposed to clear them on shutdown, so clearing them one more time is redundant and may mask logic errors. It might be worthwhile to add an assertion, though.
* shutdown process is dispatched on the queue, so the in-progress operation during the shutdown is guaranteed to be the shutdown itself;
* this only leaves any other immediate operations that might be enqueued right after shutdown completes. This behavior is not currently supported and would be a user error; it would lead to app crashing. Supporting that case is not bad, but not high-priority and can be postponed. Implementing it would be a little cumbersome because it's undesirable to wait for potentially long-running operations to finish. The way to go would be to turn those into no-ops, similar to delayed operations, which means they would have to be stored, etc.
* in short, the existing code base either should not get into problematic situations, or else not calling `clear` will not make the situation worse than what is currently in place.

File: Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm (hunk 2 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm]: You're right; I misunderstood the issue I was experiencing. The real problem was calling `dispatch_sync` on the current queue, which happened to be the main queue (from the docs, _"Calling this function and targeting the current queue results in deadlock"_). Reworked the assertion, PTAL.

File: Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm [hunk 3]
```diff
@@ -66,9 +66,8 @@ - (void)testDispatchAsyncBlocksSubmissionFromTasksOnTheQueue {
   XCTAssertNotNil(caught);
 
   XCTAssertEqualObjects(caught.name, NSInternalInconsistencyException);
-  XCTAssertTrue(
-      [caught.reason hasPrefix:@"FIRESTORE INTERNAL ASSERTION FAILED: "
-                               @"dispatchAsync called when we are already running on target"]);
+  XCTAssertTrue([caught.reason hasPrefix:@"FIRESTORE INTERNAL ASSERTION FAILED: "
+                                         @"Enforcing sequential order failed"]);
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm]: The full error message is _"Enforcing sequential order failed: currently executing operations cannot enqueue more operations "_. I just omitted a part of the assertion because I thought it's not necessary to be so specific for the test. Part of the reason I omitted `dispatchAsync` bit is because the same check is applied to `dispatchSync` -- OTOH, the latter is a second-class citizen, so perhaps it's not as important. If you think the full message is still less clear, I'll rework it.

File: Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm [hunk 4]
```diff
@@ -216,10 +216,12 @@ - (void)testEnterCheckedOperationDisallowsNesting {
 - (void)testCanScheduleCallbacksInTheFuture {
   _expectation = [self expectationWithDescription:@"Expected steps"];
   _expectedSteps = @[ @1, @2, @3, @4 ];
-  [_queue dispatchAsync:[self blockForStep:1]];
-  [_queue dispatchAfterDelay:0.005 timerID:timerID1 block:[self blockForStep:4]];
-  [_queue dispatchAfterDelay:0.001 timerID:timerID2 block:[self blockForStep:3]];
-  [_queue dispatchAsync:[self blockForStep:2]];
+  [_queue dispatchAsync:^{
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm]: For `dispatchAfterDelay`, it is necessary because it assumes it's being run on the queue. However, for the others, it's unnecessary, so I turned them back into `dispatchAsync` calls and moved them out.

File: Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm [hunk 5]
```diff
@@ -51,15 +54,32 @@ void DispatchAsync(const dispatch_queue_t queue, std::function<void()>&& work) {
   });
 }
 
-void DispatchSync(const dispatch_queue_t queue, Executor::Operation work) {
+void DispatchSync(const dispatch_queue_t queue, std::function<void()> work) {
+  FIREBASE_ASSERT_MESSAGE(
+      GetCurrentQueueLabel() != GetQueueLabel(queue),
+      "Calling dispatch_sync on the current queue will lead to a deadlock.");
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm]: Nit: `s/dispatch_sync/DispatchSync/`

File: Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm (hunk 3 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm]: I understand your motivation. Maybe something like "Enqueue methods" -- a plural that suggests more than just exactly "Enqueue" -- would be a suitable subject for a sentence more similar to the old one?

This would at least more directly suggest a binding between the action that was taken and the failure. As it is you need to understand that the AsyncQueue enforces some property and then find how your code may have interacted with that property to ultimately determine which call was problematic.

[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm]: Rephrased to be very similar to the previous version - " _Enqueue methods cannot be called_ when we are already running on target _executor_".

File: Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm (hunk 5 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm]: Done.
